File: apt/src/main/java/com/example/apt/DomFactory.java
package com.example.apt;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * Created by weihansheng.johan on 2023/1/7
 *
 * @author weihansheng.johan@bytedance.com
 */
@Retention(RetentionPolicy.CLASS) //该注解只保留到编译时
@Target(ElementType.TYPE) //该注解只作用与类、接口、枚举
public @interface DomFactory {
    String type();
    Class superClass();
}

File: app/src/main/java/com/example/hybriddemo/core/ui/dom/DomFactory.kt
package com.example.hybriddemo.core.ui.dom

import com.eclipsesource.v8.V8Object

/**
 * Created by weihansheng.johan on 2023/1/4
 * 将标签转为对应domelement
 * @author weihansheng.johan@bytedance.com
 */
//apt中实验注解生成本工厂类
object DomFactory {
    @JvmStatic
    fun create(rootObj: V8Object): DomElement? {
        val type = rootObj.getString("type")
        when (type) {
            "text" -> {
                val domText = DomText()
                domText.parse(rootObj)
                return domText
            }
            "button" -> {
                val domButton = DomButton()
                domButton.parse(rootObj)
                return domButton
            }
            "verticalLayout" -> {
                val domVerticalLayout = DomVerticalLayout()
                domVerticalLayout.parse(rootObj)
                return domVerticalLayout
            }
        }
        return null
    }
}